### PR TITLE
Convert to simd

### PR DESCRIPTION
**Pull request by ITO444**
replace instructions in the computation tree with simd instructions

### CODE REVIEW

File: backend/amd64/simd_selection.ml [hunk 1]
```diff
@@ -463,3 +463,347 @@ let () =
...
+  let create_const_vec consts =
+    let rec get_list list n =
+      if n = 0
+      then [], list
+      else
+        match list with
+        | [] -> assert false
+        | hd :: tl ->
+          let list1, list2 = get_list tl (n - 1) in
+          hd :: list1, list2
```
Comment (xclerc) [inline backend/amd64/simd_selection.ml:516]: (Isn't it `Misc.List.split_at`?)

File: backend/amd64/simd_selection.ml [hunk 2]
```diff
@@ -463,3 +463,347 @@ let () =
...
+
+type vectorized_instruction =
+  { operation : Cfg.operation;
+    arguments : register array;
+    results : register array
+  }
+
+let vector_width = 128
+
+let vectorize_operation width (cfg_ops : Cfg.operation list) :
```
Comment (xclerc) [inline backend/amd64/simd_selection.ml:487]: (I think this function is too long
to be easy to understand, but
maybe it is enough to simply
move some nested functions.)

File: backend/amd64/simd_selection.ml [hunk 3]
```diff
@@ -463,3 +463,347 @@ let () =
...
+    | Iadd -> Option.bind add_op (make_default 2 1)
+    | Isub ->
+      let sse_op =
+        match width_type with
+        | W64 -> Sub_i64
+        | W32 -> Sub_i32
+        | W16 -> Sub_i16
+        | W8 -> Sub_i8
+      in
+      Cfg.Specific (Isimd (SSE2 sse_op)) |> make_default 2 1
```
Comment (xclerc) [inline backend/amd64/simd_selection.ml:557]: Why is this case different from
the ones for `Iadd` / `Imul`?

Comment (xclerc) [reply backend/amd64/simd_selection.ml:557]: (asking because if we can make
the cases look similar, it will make
review an maintenance easier.)

**Review by xclerc — COMMENTED** [review]
(Only a fairly superficial first pass.)

File: backend/amd64/simd_selection.ml (hunk 3 above)
Comment (ITO444) [reply backend/amd64/simd_selection.ml:557]: oh, add and mul are reused in lea but not the rest of the functions, should I still pull out all the other instructions like add and mul?

File: backend/cfg/vectorize.ml [hunk 4]
```diff
@@ -178,85 +511,1048 @@ end = struct
...
+    match cell_option with
+    | None -> ()
+    | Some cell ->
+      let instructon = Instruction.Basic (DLL.value cell) in
+      let next_cell = DLL.next cell in
+      if Instruction.Id.Set.exists
+           (Instruction.Id.equal (Instruction.id instructon))
+           good_instructions
+      then DLL.delete_curr cell;
+      remove_scalar_instructions next_cell
```
[the remark this answers is not in the document]
Comment (ITO444) [reply backend/cfg/vectorize.ml:1509]: done

File: backend/amd64/simd_selection.ml [hunk 5]
```diff
@@ -463,3 +463,347 @@ let () =
...
+  | W8
+  | W16
+  | W32
+  | W64
+
+type register =
+  | New of int
+  | Argument of int
+  | Result of int
+  | Original of int
```
[the remark this answers is not in the document]
Comment (ITO444) [reply backend/amd64/simd_selection.ml:477]: added comments

File: backend/amd64/simd_selection.ml (hunk 2 above)
Comment (ITO444) [reply backend/amd64/simd_selection.ml:487]: what nested functions do you suggest to move out?

File: backend/amd64/simd_selection.ml (hunk 1 above)
Comment (ITO444) [reply backend/amd64/simd_selection.ml:524]: I can't seem to find that

Comment (ITO444) [reply backend/amd64/simd_selection.ml:524]: you mean Misc.split_at?

Comment (ITO444) [reply backend/amd64/simd_selection.ml:524]: I seem to not be able to call that either

Comment (ITO444) [reply backend/amd64/simd_selection.ml:524]: ah it's Misc.Stdlib.List.split_at

**Comment by gretay-js** [comment]
superseded by https://github.com/ocaml-flambda/flambda-backend/pull/3365

**Pull request closed without merging**